packages/playwright-core/src/server/injected/roleUtils.ts

```diff
@@ -1005,13 +1005,14 @@ function isNativelyDisabled(element: Element) {
   return isNativeFormControl && (element.hasAttribute('disabled') || belongsToDisabledFieldSet(element));
 }
 
-function belongsToDisabledFieldSet(element: Element | null): boolean {
-  if (!element)
+function belongsToDisabledFieldSet(element: Element): boolean {
+  const fieldSetElement = element?.closest('FIELDSET[DISABLED]');
+  if (!fieldSetElement)
     return false;
-  if (elementSafeTagName(element) === 'FIELDSET' && element.hasAttribute('disabled'))
+  const legendElement = fieldSetElement.querySelector('LEGEND');
```

Comment:
To only match immediate children, which should be faster:

```suggestion
  const legendElement = fieldSetElement.querySelector(':scope > LEGEND');
```